temporal-sdk/src/main/java/io/temporal/internal/common/SearchAttributePayloadConverter.java

```diff
@@ -0,0 +1,254 @@
+            ex);
+      }
+    }
+  }
+
+  private boolean isUnset(@Nonnull Payload payload) {
+    try {
+      List<?> o = defaultDataConverter.fromPayload(payload, List.class, List.class);
+      if (o.size() == 0) {
+        // this is an "unset" token, we don't need a type for it
+        return true;
+      }
+    } catch (Exception e) {
+      // ignore the exception, it was an attempt to parse a specific "unset" ('[]') value only
+    }
+    return false;
+  }
+
+  private static IndexedValueType getIndexType(ByteString dataType) {
+    if (dataType != null) {
+      String dataTypeString = dataType.toStringUtf8();
+      if (dataTypeString.length() != 0) {
+        return encodedValueToIndexValueType(dataTypeString);
+      }
+    }
+    return null;
+  }
+
+  @Nullable
+  private static String extractIndexValueTypeName(@Nonnull Object instance) {
+    IndexedValueType type;
+    if (instance instanceof Collection) {
+      Collection<?> collection = (Collection<?>) instance;
+      if (!collection.isEmpty()) {
+        List<IndexedValueType> indexValues =
+            collection.stream()
+                .map(k -> (javaTypeToIndexValueType(k.getClass())))
+                .distinct()
+                .collect(Collectors.toList());
+
+        if (indexValues.size() == 1) {
+          type = indexValues.get(0);
+        } else {
+          throw new IllegalArgumentException(
+              instance + " maps into a mix of IndexValueTypes: " + indexValues);
+        }
+      } else {
+        // it's an "unset" value
+        // has to be null and can't be INDEXED_VALUE_TYPE_UNSPECIFIED
+        // because there was a bug: https://github.com/temporalio/temporal/issues/2693
+        return null;
+      }
+    } else {
+      type = javaTypeToIndexValueType(instance.getClass());
+    }
+    //  We can't enforce the specific types because of backwards compatibility. Previously any value
+    // that was serializable by json into a string could be passed as a search attribute
+    //    if (isIndexTypeUndefined(type)) {
+    //      throw new IllegalArgumentException("Instance " + instance + " of class " +
+    // instance.getClass() + " is not supported as a search attribute value");
+    //    }
+    return indexValueTypeToEncodedValue(type);
+  }
+
+  @Nonnull
+  private static IndexedValueType javaTypeToIndexValueType(@Nonnull Class<?> type) {
+    if (CharSequence.class.isAssignableFrom(type)) {
+      return IndexedValueType.INDEXED_VALUE_TYPE_TEXT;
+    } else if (Long.class.equals(type)
+        || Integer.class.equals(type)
+        || Short.class.equals(type)
+        || Byte.class.equals(type)) {
+      return IndexedValueType.INDEXED_VALUE_TYPE_INT;
+    } else if (Double.class.equals(type) || Float.class.equals(type)) {
+      return IndexedValueType.INDEXED_VALUE_TYPE_DOUBLE;
+    } else if (Boolean.class.equals(type)) {
+      return IndexedValueType.INDEXED_VALUE_TYPE_BOOL;
+    } else if (OffsetDateTime.class.equals(type)) {
```

Comment:
Hrmm, I wonder if `Instant` or even `java.util.Date` should be supported? Probably not if we expect to be able to decode to the same type we receive.

Reply:
Date is, probably, not doable looking at the underlying ES Type.
 
Instant is a really good thought though, but yes, to your point, if we do Instant, we will have to return Instant ideally. And it will restrict us because OffsetDateTime is a more powerful class.